File: 프로그래머스/lv2/87946. 피로도/피로도.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

int solution(int k, vector<vector<int>> dungeons) {
    int answer = -1;
    
    sort(dungeons.begin(), dungeons.end());
    
    do{
        int count = 0;
        int current_k = k;
        for(auto& i : dungeons){
            if(current_k >= i[0]){
                current_k -= i[1];
                count++;
            }
        }
        answer = max(answer, count);
    }while (next_permutation(dungeons.begin(), dungeons.end()));
    
    return answer;
}

File: 프로그래머스/2/87946. 피로도/피로도.py
def solution(k, dungeons):
    def backtracking(d, k):
        nonlocal answer
        
        can = False
        
        for i in range(len(dungeons)):
            if not visited[i] and dungeons[i][0] <= k:
                can = True
                visited[i] = 1
                backtracking(d+1, k-dungeons[i][1])
                visited[i] = 0
        
        if not can:
            answer = max(d, answer)
            return 
                
    
    answer = -1
    visited = [0] * len(dungeons)
    backtracking(0, k)
    
    return answer